File: Algorithm/UnionFind/programmers_네트워크.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int Find(vector<int> &parents, int x) {
    if (x == parents[x])
        return x;
    return x = Find(parents, parents[x]);
}

void Union(vector<int> &parents, int x, int y) {
    x = parents[x];
    y = parents[y];
    
    if (x != y) {
        if (x < y)
            parents[y] = x;
        else
            parents[x] = y;
    }
    
    return;
}

int solution(int n, vector<vector<int>> computers) {
    vector<int> answer;
    vector<int> parents(n+1, 0);
    
    for (int i = 0; i < n; ++i)
        parents[i] = i;
    
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            if (i == j || !computers[i][j])
                continue;
            
            Union(parents, i, j);
        }
    }
    
    for (int i = 0; i < n; ++i) {
        if (find(answer.begin(), answer.end(), Find(parents, i)) == answer.end())
            answer.push_back(i);
    }
    
    return answer.size();
}

File: Algorithm/UnionFind/programmers_네트워크.swift
import Foundation

func solution(_ n:Int, _ computers:[[Int]]) -> Int {
    var answer:[Int] = []
    var parents: [Int] = Array(repeating: 0, count: n+1)
    
    func Find(_ x : Int) -> Int {
        if (x != parents[x]) {
            parents[x] = Find(parents[x])
        }
        return parents[x]
    }
    
    func Union(_ x: Int, _ y: Int) {
        var x = x
        var y = y
        x = Find(x)
        y = Find(y)
        
        if x != y {
            if x < y {
                parents[y] = x
            } else {
                parents[x] = y
            }
        }
        
        return
    }
    
    for i in 0...n {
        parents[i] = i
    }
    
    for i in 0..<n {
        for j in 0..<n {
            if (i == j || computers[i][j] == 0) {
                continue
            }
            
            Union(i, j)
        }
    }
    
    for i in 0..<n {
        if !answer.contains(Find(i)) {
            answer.append(i)
        }
    }
    return answer.count
}